wave_lang/kernel/wave/analysis/index_sequence_analysis.py

```diff
@@ -946,26 +949,109 @@ def apply_offset(node: fx.Node):
     """
-    with binary_op.graph.inserting_before(binary_op.fx_node):
+    to_broadcast = get_custom(getattr(op, to_broadcast_arg_name))
+
+    with op.graph.inserting_before(op.fx_node):
         broadcasted = Broadcast(
             to_broadcast.fx_node, target_node.type.symbolic_shape
-        ).add_to_graph(binary_op.graph)
+        ).add_to_graph(op.graph)
         custom = get_custom(broadcasted)
-        custom.vector_shapes = binary_op.vector_shapes
+        custom.vector_shapes = op.vector_shapes
         custom.index = deepcopy(target_node.index)
         custom.index[broadcast_dim].size = broadcast_size
-        broadcast_idx = list(binary_op.node_args.values()).index(to_broadcast)
-        binary_op.update_arg(broadcast_idx, custom.fx_node)
+
+        op.update_arg(to_broadcast_arg_name, custom.fx_node)
+
+
+def resolve_broadcasting_for_op(
+    trace: CapturedTrace, op_type: type, operand_identifiers: list[str]
+):
+    """Generic broadcasting resolution for any multi-operand operation."""
+
+    ops = trace.walk(lambda node: isinstance(get_custom(node), op_type))
+
+    for op_node in ops:
+        custom = get_custom(op_node)
+
+        operands = []
+        for identifier in operand_identifiers:
+            operand_custom = get_custom(getattr(custom, identifier))
+
+            index = get_index(operand_custom)
+            dim, size = get_largest_index_and_size(index, operand_custom)
+            operands.append(
+                {"custom": operand_custom, "dim": dim, "size": size, "id": identifier}
+            )
+
+        sizes = [operand["size"] for operand in operands]
+        if len(set(sizes)) <= 1 or all(size == 1 for size in sizes):
+            continue
+
+        target = max(operands, key=lambda x: x["size"])
+
+        def generate_error_context():
+            context_lines = [f"\n{op_type.__name__.lower()}={custom}"]
+            for op in operands:
+                context_lines.extend(
+                    [
+                        f"\n{op['id']}: {op['custom']}",
+                        f"\n{op['id']}_index: {get_index(op['custom'])}",
+                        f"\n{op['id']}_dim: {op['dim']}",
+                        f"\n{op['id']}_size: {op['size']}",
+                        f"\n{op['id']}.type.symbolic_shape: {op['custom'].type.symbolic_shape}",
+                    ]
+                )
+            return "".join(context_lines)
+
+        for operand in operands:
+            if operand["size"] < target["size"]:
+                if operand["size"] > 1 and target["size"] > 1:
+                    raise NotImplementedError(
+                        f"Currently only support resolving discrepancies when one of the shapes is 1."
+                        f"{generate_error_context()}"
+                    )
+
+                if operand["dim"] != target["dim"]:
+                    missing_dims = set(target["custom"].type.symbolic_shape).difference(
```

Comment:
Can we add back the comment
```
            # If the dimensions don't agree, we can still do this broadcast only if
            # the two nodes differ in shape along the broadcasting dimension and the
            # broadcasting dimension is the innermost dimension.
```

For better context for the reader on what's happening